### DISCUSSION TITLE
Support Yarn 3 workspaces

### DISCUSSION THREAD

**Discussion by eric-burel** [remix-run/remix#3668] [Proposals]
Hey folks, I maintain a stack in a monorepo (https://github.com/VulcanJS/vulcan-npm) that contains a Next, an Express and a Remix boilerplate alongside multiple packages.

The issue I have, is that if you define your Remix app as an independant workspace, it cannot have a `yarn.lock` file, otherwise it disable Yarn package hoisting (which is the whole point of having a monorepo in the first place).

You could simply run `yarn` in your Remix folder, but then versions will be inconsistant, it's not acting as a lock anymore.

So you need to find a way to generate a `yarn.remix.lock` based on the big, root `yarn.lock` which will include all dependencies of your app.

- I've added documentation to this Yarn plugin that can generate `yarn.lock` for workspaces of a monorepo: https://github.com/JanVoracek/yarn-plugin-entrypoint-lockfiles
  It will read your big root `yarn.lock`, your Remix app `package.json`, and generate a `yarn.remix.lock` with the right packages

- I've commented on the parent issue of Yarn (which doesn't support this out of the box sadly) https://github.com/yarnpkg/berry/issues/1223#issuecomment-1177250693

- This PR demoes our implementation in Vulcan: https://github.com/VulcanJS/vulcan-npm/pull/132

It extends to other frameworks as well. I could finally run Remix CI actions (the ones defined in the Indie stack) correctly with this approach and a few tweaks.

It's very far-fetched, but needed if you want to publish a Remix stack from a monorepo.

Related: https://github.com/remix-run/remix/issues/2474 for NPM workspaces, https://github.com/remix-run/remix/issues/683 for yarn 3 bugs, PR to support pnp https://github.com/remix-run/remix/pull/1316

**Comment by HosseinAgha** [comment]
@eric-burel Could you please explain why you need `yarn.lock` file for each workspace? Do you need it for deployment and production?

**Reply by eric-burel** [reply]
I need it for the applications only, not necessarily each workspace. This is needed because I then publish the starter app as an isolated read-only repository, that end users may clone. This repo needs its own yarn.lock, so that end user get the right package versions and it's also useful for CI (I could run the CI in the monorepo though, but it's nice to keep it alongside the boilerplate code because people can tweak it after cloning/forking).
